sdks/java/io/synthetic/src/main/java/org/apache/beam/sdk/io/synthetic/SyntheticStep.java

```diff
@@ -0,0 +1,203 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.sdk.io.synthetic;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.RateLimiter;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.metrics.Counter;
+import org.apache.beam.sdk.metrics.Metrics;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+import org.joda.time.Duration;
+
+/**
+ * This {@link SyntheticStep} class provides a parameterizable {@code DoFn} that consumes {@code
+ * KV<byte[], byte[]>} and emits {@code KV<byte[], byte[]>}. Each emitted record will be slowed down
+ * by a certain sleep time generated based on the specified sleep time distribution in {@link
+ * SyntheticStep.Options}.
+ *
+ * <p>See {@link SyntheticStep.Options} for how to construct an instance. To construct a {@code
+ * SyntheticStep} from SyntheticStep options, use:
+ *
+ * <pre>{@code
+ * SyntheticStep.Options options = ...;
+ *
+ * // Construct the synthetic step with options.
+ * SyntheticStep step = new SyntheticStep(options);
+ * }</pre>
+ */
+public class SyntheticStep extends DoFn<KV<byte[], byte[]>, KV<byte[], byte[]>> {
+  private final Options options;
+  private final KV<Long, Long> idAndThroughput; // used when maxWorkerThroughput is set
+  private final Counter throttlingMsecsCounter =
+      Metrics.counter("dataflow-throttling-metrics", "throttling-msecs");
+
+  public SyntheticStep(Options options) {
+    options.validate();
+    this.options = options;
+    // use a random id so that a pipeline could have multiple SyntheticSteps
+    this.idAndThroughput = KV.of(new Random().nextLong(), options.maxWorkerThroughput);
+  }
+
+  private KV<byte[], byte[]> outputElement(
+      byte[] inputKey, byte[] inputVal, long hashCodeOfVal, int index, Random rnd) {
+    long seed = options.hashFunction.hashLong(hashCodeOfVal + index).asLong();
+    Duration delayMsec = Duration.millis(options.nextDelay(seed));
+    long sleepMillis = 0;
+    while (delayMsec.getMillis() > 0) {
+      sleepMillis +=
+          SyntheticUtils.delay(
+              delayMsec, options.cpuUtilizationInMixedDelay, options.delayType, rnd);
+
+      // Ensure the worker throughput is within the limit (if set).
+      if (options.maxWorkerThroughput < 0
```

Comment:
We could extract a method out of this instead of writing the comment: 

```
  private boolean isWithinTheThroughputLimit() {
    return options.maxWorkerThroughput < 0 || rateLimiterCache.getUnchecked(idAndThroughput).tryAcquire();
  }
```

Reply:
That makes sense. Done.